File: src/main/kotlin/dev/blackoutburst/game/texture/Texture.kt
package dev.blackoutburst.game.texture

import dev.blackoutburst.game.utils.IOUtils
import dev.blackoutburst.game.utils.stack
import org.lwjgl.opengl.GL30.*
import org.lwjgl.stb.STBImage

class Texture(filePath: String) {
    var id = -1

    init {
        stack { stack ->
            id = glGenTextures()
            glBindTexture(GL_TEXTURE_2D, id)

            val comp = stack.mallocInt(1)
            val width = stack.mallocInt(1)
            val height = stack.mallocInt(1)

            STBImage.stbi_set_flip_vertically_on_load(true)

            val data = STBImage.stbi_load_from_memory(
                IOUtils.ioResourceToByteBuffer(filePath, 4096),
                width,
                height,
                comp,
                4
            )

            glTexImage2D(
                GL_TEXTURE_2D,
                0,
                GL_RGBA,
                width[0],
                height[0],
                0,
                GL_RGBA,
                GL_UNSIGNED_BYTE,
                data
            )

            comp.clear()
            data?.let { STBImage.stbi_image_free(it) }

            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST)
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST)
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT)
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT)

        }
    }
}

File: C/src/graphics/texture.c
#define GLFW_INCLUDE_NONE
#define STB_IMAGE_IMPLEMENTATION

#include <stdio.h>
#include <stdlib.h>
#include "utils/types.h"
#include "glfw/glfw3.h"
#include "graphics/opengl.h"
#include "utils/math.h"
#include "stb/stb_image.h"

void textureDelete(U32 id) {
    glDeleteTextures(1, &id);
}

U32 textureCreate(const I8* path, U8 flip) {
    U32 id = 0;
    glGenTextures(1, &id);

    glBindTexture(GL_TEXTURE_2D, id);

    I32 w = 0;
    I32 h = 0;
    I32 comp = 0;

    stbi_set_flip_vertically_on_load(flip);
    
    U8* data = stbi_load(path, &w, &h, &comp, 4);

    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, w, h, 0, GL_RGBA, GL_UNSIGNED_BYTE, data);

    stbi_image_free(data);
    
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);

    stbi_set_flip_vertically_on_load(0);

    return id;
}
